### DISCUSSION TITLE
modify source model from edit-label-handler

### DISCUSSION THREAD

**Discussion by rileydanejohnston** [eclipse-glsp/glsp#769] [General] [answered]
Hello,
I’m trying to get a better understanding of the edit-label-handler so I can modify my source model within the edit-label-handler. My source model has Node objects that have a ‘name’ property. Nodes end up rendering a GNode with a GLabel child, where the ‘name’ property becomes the label text. I’d like to modify the name property in my source model within the edit-label-handler. Here’s my implementation of the edit-label-handler:

```
 import { ApplyLabelEditOperation } from '@eclipse-glsp/protocol';
 import { GLSPServerError, OperationHandler, toTypeGuard, GNode } from '@eclipse-glsp/server-node';
 import { inject, injectable } from 'inversify';
 import { SourceModelState } from '../model/model-state';

 @injectable()
 export class EditLabelHandler implements OperationHandler {
	readonly operationType = ApplyLabelEditOperation.KIND;

	@inject(SourceModelState)
	protected modelState: SourceModelState;

	execute(operation: ApplyLabelEditOperation): void {
		const index = this.modelState.index;
		const allSourceNodes = this.modelState.sourceNodes;

		// Retrieve the parent node of the label that should be edited
		const indexNode = index.findParentElement(operation.labelId, toTypeGuard(GNode));
		if (indexNode) {
			// find source model match
			const sourceNode = allSourceNodes.find(node => node.id === indexNode.id);
			if (sourceNode) {
				// update source model with new label text
				sourceNode.name = operation.text;
			}
		} else {
			throw new GLSPServerError(`Could not retrieve the parent task for the label with id ${operation.labelId}`);
		}
	}
 }
```

The GLabel text is updating correctly when I console log it. The GLabel gets updated no matter what is in the edit-label-handler’s execute function (even a console.log with  nothing else), which is confusing to me. I have a workaround where I extract the updated text and update the source model in the gmodel-factory, but I imagine I should be able to do this in the edit-label-handler. Is it possible to edit the source model from within the edit-label-handler? Am I missing something?

Other details:
- I’m using a node backend with a vscode client.
- I’ve binded the handler & validator in the module file
- editLabelValidator is firing and returning status 4
- GLabels/GNodes in the graph are editable

**Comment by rileydanejohnston** [comment, marked as answer]
I've found a solution for this. I went in diagram-module file on the server side and updated the order of the bindings so that the super.configureOperationHandlers(binding) line is last, with my custom handlers coming first like so:
```
protected override configureOperationHandlers(binding: InstanceMultiBinding<OperationHandlerConstructor>): void {
   binding.add(handler1);
   binding.add(handler2);
   super.configureOperationHandlers(binding);
}
```

**Reply by planger** [reply]
Thanks for posting your solution!

Are you extending the `[GModelDiagramModule](https://github.com/eclipse-glsp/glsp-server-node/blob/main/packages/server-node/src/gmodel-lib/gmodel-diagram-module.ts)`? In this module, the `ApplyLabelEditOperationHandler` is added. Afaik for operation handlers, the first matching handler is executed. Thus, it would be probably more clean, if you'd remove the `ApplyLabelEditOperationHandler` and add your handler instead.
```typescript
protected override configureOperationHandlers(binding: InstanceMultiBinding<OperationHandlerConstructor>): void {
    super.configureOperationHandlers(binding);
    binding.remove(ApplyLabelEditOperationHandler);
    binding.add(handler1);
    binding.add(handler2);
}
```
